### PR TITLE
feat(ASC-262): multi cluster deployment

### PR DESCRIPTION
**Pull request by gbenhaim**
This change introduce the ability to deploy StoneSoup into multiple clusters. The general guideline is to have reduce as much as possible the scripts that are needed for deployment, and relay on ArgoCD instead.

The change can be split into 4 topics:

ArgoCD:

Instead of using Applications, use ApplicationSets with the cluster generator the is configured with label selectors (using Kustomize).

Compose the different ApplicationSets in directories that reflect the role of the cluster in the fleet (e.g host, member).

Hack Directory:

Move the logic that responsible for creating secrets into dedicated scripts.

Create bootstrap scripts for the different cluster types.

Adjust preview.sh to work with the new ArgoCD manifests.

Move the ArgoCD deployment to dedicated script.

Note - This change preserves the ability to install StoneSoup into a single cluster for development purposes.

Components Directory:

The Gitops service requires that the openshift-gitops operator will be installed in the cluster. Deploy the subscription as part of the deployment of the Gitops service.

Docs:

Update the docs according to the changes. The multi cluster docs are still not complete. Those can be completed when we sort out the injection of secrets to the clusters.

Out of scope:

- Vault deployment in member clusters
- Injecting the required secrets for the StoneSoup components from a central location.

Signed-off-by: gbenhaim <gbenhaim@redhat.com>

### CODE REVIEW

**Comment by Dannyb48** [comment]
@gbenhaim this looks to failed during deployment, the registration service failed to start. One of the container is in a crashloop. Since you are working refactoring the deployment as part of PR I did not file a bug.

**Comment by gbenhaim** [comment]
thanks @Dannyb48, I fixed the problem.

File: argo-cd-apps/base/member/infra-deployments/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+  - gitops
```
Comment (Michkov) [inline argo-cd-apps/base/member/infra-deployments/kustomization.yaml:4]: I would prefer to point directly to gitops.yaml which could be in argo-cd-apps/base/member/infra-deployments - avoiding creation folder with another kustomization.yaml (which can be created later if needed.)

**Review by Michkov — COMMENTED** [review]
Two minor comments. Only having issue with unstable argocd route -> when you run './hack/preview.sh' then the argoCD web interface is not available for some time.

**Review by Michkov — APPROVED** [review]

File: argo-cd-apps/base/member/infra-deployments/kustomization.yaml (hunk 1 above)
Comment (gbenhaim) [reply argo-cd-apps/base/member/infra-deployments/kustomization.yaml:4]: The purpose of putting it and the other apps in their own folder is to allow picking which to use for each environment. Kustomize doesn't let you specify a specific yaml as a base, just Kustomize directories. At this point where everything changes rapidly, I would like to keep this flexibility.

File: hack/secret-creator/create-github-secret.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -e
+
+main() {
+    local github_token=${1:?"Github toekn was not provided"}
```
[the remark this answers is not in the document]
Comment (gbenhaim) [reply hack/secret-creator/create-github-secret.sh]: done

File: argo-cd-apps/base/local-cluster-secret/local-cluster-secret.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+kind: Secret
+metadata:
+  labels:
+    argocd.argoproj.io/secret-type: cluster
+  name: cluster-kubernetes.default.svc
+type: Opaque
+data:
+  config: eyJ0bHNDbGllbnRDb25maWciOnsiaW5zZWN1cmUiOmZhbHNlfX0=
+  name: aW4tY2x1c3Rlci1sb2NhbA==
+  server: aHR0cHM6Ly9rdWJlcm5ldGVzLmRlZmF1bHQuc3Zj
```
[the remark this answers is not in the document]
Comment (gbenhaim) [reply argo-cd-apps/base/local-cluster-secret/local-cluster-secret.yaml]: done

**Comment by gbenhaim** [comment]
Me and @Michkov agreed that it would be better to push this change to the `multi-cluster` branch and use it for deploying the new staging instance. Once it's done, and the new instance is functional, we can proceed with this PR.

**Comment by gbenhaim** [comment]
replaced with https://github.com/redhat-appstudio/infra-deployments/pull/1247

**Pull request closed without merging**